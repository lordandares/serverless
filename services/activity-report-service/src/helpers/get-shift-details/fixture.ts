export const application = {
  _id: '565fe59257c5d59e60cc35b1',
  name: 'Testing Application',
}

export const user = {
  application: '565fe59257c5d59e60cc35b5',
  email: 'testing@test.com',
  firstName: 'Testing',
  preferences: {
    notifications: {
      channels: [
        {
          enabled: true,
          name: 'daily-location-report',
          options: {
            hours: 9,
            timezone: 'America/Los_Angeles',
          },
        },
      ],
    },
  },
  permissions: [
    {
      module: 'area',
      type: 'document',
      value: '565fe59257c5d59e60cc35b3',
    },
    {
      module: 'area',
      type: 'document',
      value: '565fe59257c5d59e60cc35b4',
    },
  ],
}

export const legacyUser = {
  application: '565fe59257c5d59e60cc35b5',
  email: 'testing@test.com',
  firstName: 'Testing',
  preferences: {
    notifications: {
      channels: [
        {
          enabled: true,
          name: 'daily-location-report',
          options: {
            hours: 9,
            timezone: 'America/Los_Angeles',
          },
        },
      ],
    },
  },
  permissions: [
    {
      module: 'location',
      type: 'document',
      value: '565fe59257c5d59e60cc35b3',
    },
  ],
}

export const userWithNoPermissions = {
  application: '565fe59257c5d59e60cc35b8',
  email: 'testing@test.com',
  firstName: 'Testing',
  preferences: {
    notifications: {
      channels: [
        {
          enabled: true,
          name: 'daily-location-report',
          options: {
            hours: 9,
            timezone: 'America/Los_Angeles',
          },
        },
      ],
    },
  },
  permissions: [],
}

export const locations = [
  {
    _id: '565fe59257c5d59e60cc35b3',
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    name: 'Testing Location 1',
  },
  {
    _id: '565fe59257c5d59e60cc35b4',
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b4/1999/12/1999-12-30-1548194730622.pdf',
        timestamp: new Date('2019-12-30T18:00:00Z'),
      },
    },
    name: 'Testing Location 2',
  },
  {
    _id: '565fe59257c5d59e60cc35b5',
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b5/1999/12/1999-12-29-1548194730622.pdf',
        timestamp: new Date('2019-12-29T18:00:00Z'),
      },
    },
    name: 'Testing Location 3',
    properties: {
      area: '565fe59257c5d59e60cc35c1',
    },
  },
]

export const shifts = [
  {
    _id: '5b480e6120d8f51322b0bfee',
    application: '565fe59257c5d59e60cc3123',
    user: '565fe59257c5d59e60cc0001',
    start: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b3',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b2',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    location: null,
    duration: 9,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
  {
    _id: '60082e79dc704419ac763517',
    user: '565fe59257c5d59e60cc0002',
    application: '565fe59257c5d59e60cc3123',
    start: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b0',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b0',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    location: null,
    duration: 6,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
  {
    _id: '60082e79dc704419ac763123',
    user: '565fe59257c5d59e60cc0003',
    application: '565fe59257c5d59e60cc35b5',
    start: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b2',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      area: {
        location: {
          id: '565fe59257c5d59e60cc35b3',
        },
      },
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    location: null,
    duration: 9,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
]

export const legacyShifts = [
  {
    _id: '5b480e6120d8f51322b0bfed',
    application: '565fe59257c5d59e60cc3123',
    location: '565fe59257c5d59e60cc35b3',
    user: '565fe59257c5d59e60cc0001',
    start: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    duration: 9,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
  {
    _id: '60082e79dc704419ac763518',
    user: '565fe59257c5d59e60cc0002',
    application: '565fe59257c5d59e60cc3123',
    location: '565fe59257c5d59e60cc35b3',
    start: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    duration: 6,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
  {
    _id: '60082e79dc704419ac763124',
    user: '565fe59257c5d59e60cc0003',
    location: '565fe59257c5d59e60cc35b4',
    application: '565fe59257c5d59e60cc35b5',
    start: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    end: {
      time: new Date('2019-12-30T18:00:00Z'),
    },
    files: {
      pdf: {
        path:
          'activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf',
        timestamp: new Date('2019-12-31T18:00:00Z'),
      },
    },
    duration: 9,
    createdAt: new Date('2019-12-30T18:00:00Z'),
  },
]
