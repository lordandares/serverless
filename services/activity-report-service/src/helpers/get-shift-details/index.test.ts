const mockToArray = jest.fn()
const mockSort = jest.fn().mockReturnValue({ toArray: mockToArray })
const mockFind = jest.fn().mockReturnValue({ sort: mockSort })
const mockFindOne = jest.fn()

const mockGetCollection = jest.fn().mockResolvedValue({
  find: mockFind,
  findOne: mockFindOne,
})

import { mongo } from '@lighthouse/serverless-common'

import { getShiftDetails } from '.'
import * as mockData from './fixture'

mongo.getCollection = mockGetCollection

describe('helpers:getShiftDetails', () => {
  afterEach(() => jest.clearAllMocks())

  it('it returns filtered shifts when there are area location permissions', async () => {
    expect.assertions(6)
    const datetime = new Date('2019-11-07T11:15:00Z')

    mockToArray.mockResolvedValue(mockData.shifts)
    mockFindOne.mockResolvedValue(mockData.application)

    const result = await getShiftDetails({ user: mockData.user, datetime })

    expect(mockGetCollection).toHaveBeenCalledTimes(2)
    expect(mockGetCollection).toBeCalledWith('shifts')
    expect(mockGetCollection).toBeCalledWith('applications')

    // shift query
    expect(mockFind).toBeCalledWith({
      application: new mongo.ObjectId(mockData.user.application),
      'end.time': {
        $gte: new Date('2019-11-05T11:15:00Z'),
        $lt: expect.any(Date),
      },
      deleted: { $ne: true },
    })
    expect(mockSort).toBeCalledWith({ 'end.time': -1 })

    expect(result).toMatchSnapshot()
  })

  it('it returns filtered shifts when there are legacy location permissions', async () => {
    expect.assertions(6)

    const datetime = new Date('2019-11-07T11:15:00Z')
    mockFindOne.mockResolvedValue(mockData.application)
    mockToArray.mockResolvedValue(mockData.legacyShifts)

    const result = await getShiftDetails({
      user: mockData.legacyUser,
      datetime,
    })

    expect(mockGetCollection).toHaveBeenCalledTimes(2)
    expect(mockGetCollection).toBeCalledWith('shifts')
    expect(mockGetCollection).toBeCalledWith('applications')

    // shift query
    expect(mockFind).toBeCalledWith({
      application: new mongo.ObjectId(mockData.legacyUser.application),
      'end.time': {
        $gte: new Date('2019-11-05T11:15:00Z'),
        $lt: expect.any(Date),
      },
      deleted: { $ne: true },
    })
    expect(mockSort).toBeCalledWith({ 'end.time': -1 })

    expect(result).toMatchSnapshot()
  })

  it('it returns all shifts when the user has no location permissions', async () => {
    expect.assertions(6)

    const datetime = new Date('2019-11-05T11:15:00Z')
    mockFindOne.mockResolvedValue(mockData.application)
    mockToArray.mockResolvedValue(mockData.shifts)

    const result = await getShiftDetails({
      datetime,
      user: mockData.userWithNoPermissions,
    })

    expect(mockGetCollection).toHaveBeenCalledTimes(2)
    expect(mockGetCollection).toBeCalledWith('shifts')
    expect(mockGetCollection).toBeCalledWith('applications')

    // shift query
    expect(mockFind).toBeCalledWith({
      application: new mongo.ObjectId(
        mockData.userWithNoPermissions.application,
      ),
      'end.time': {
        $gte: new Date('2019-11-03T11:15:00Z'),
        $lt: expect.any(Date),
      },
      deleted: { $ne: true },
    })
    expect(mockSort).toBeCalledWith({ 'end.time': -1 })

    expect(result).toMatchSnapshot()
  })

  it('it returns all locations when the legacy user has no location permissions', async () => {
    expect.assertions(6)

    const datetime = new Date('2019-11-07T11:15:00Z')
    mockFindOne.mockResolvedValue(mockData.application)
    mockToArray.mockResolvedValue(mockData.legacyShifts)

    const result = await getShiftDetails({
      datetime,
      user: mockData.userWithNoPermissions,
    })

    expect(mockGetCollection).toHaveBeenCalledTimes(2)
    expect(mockGetCollection).toBeCalledWith('shifts')
    expect(mockGetCollection).toBeCalledWith('applications')

    // shift query
    expect(mockFind).toBeCalledWith({
      application: new mongo.ObjectId(
        mockData.userWithNoPermissions.application,
      ),
      'end.time': {
        $gte: new Date('2019-11-05T11:15:00Z'),
        $lt: expect.any(Date),
      },
      deleted: { $ne: true },
    })
    expect(mockSort).toBeCalledWith({ 'end.time': -1 })

    expect(result).toMatchSnapshot()
  })

  it('handles error when returning shifts and returns false', async () => {
    expect.assertions(1)

    const error = new Error('Location error')
    const datetime = new Date('2019-11-07T11:15:00Z')

    mockFindOne.mockResolvedValue(mockData.application)
    mockToArray.mockRejectedValue(error)

    const user = {
      ...mockData.user,
      application: '565fe59257c5d59e60cc35d2',
    }

    const result = await getShiftDetails({
      datetime,
      user,
    })

    expect(result).toEqual(false)
  })
})
