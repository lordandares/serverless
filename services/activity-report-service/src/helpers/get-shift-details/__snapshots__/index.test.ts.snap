// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers:getShiftDetails it returns all locations when the legacy user has no location permissions 1`] = `
Object {
  "applicationName": "Testing Application",
  "email": "testing@test.com",
  "firstName": "Testing",
  "lastName": undefined,
  "preferences": Object {
    "notifications": Object {
      "channels": Array [
        Object {
          "enabled": true,
          "name": "daily-location-report",
          "options": Object {
            "hours": 9,
            "timezone": "America/Los_Angeles",
          },
        },
      ],
    },
  },
  "shifts": Array [
    Object {
      "_id": "5b480e6120d8f51322b0bfed",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0001",
    },
    Object {
      "_id": "60082e79dc704419ac763518",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0002",
    },
    Object {
      "_id": "60082e79dc704419ac763124",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0003",
    },
  ],
  "timezone": "America/Los_Angeles",
}
`;

exports[`helpers:getShiftDetails it returns all shifts when the user has no location permissions 1`] = `
Object {
  "applicationName": "Testing Application",
  "email": "testing@test.com",
  "firstName": "Testing",
  "lastName": undefined,
  "preferences": Object {
    "notifications": Object {
      "channels": Array [
        Object {
          "enabled": true,
          "name": "daily-location-report",
          "options": Object {
            "hours": 9,
            "timezone": "America/Los_Angeles",
          },
        },
      ],
    },
  },
  "shifts": Array [
    Object {
      "_id": "5b480e6120d8f51322b0bfee",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0001",
    },
    Object {
      "_id": "60082e79dc704419ac763517",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0002",
    },
    Object {
      "_id": "60082e79dc704419ac763123",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0003",
    },
  ],
  "timezone": "America/Los_Angeles",
}
`;

exports[`helpers:getShiftDetails it returns filtered shifts when there are area location permissions 1`] = `
Object {
  "applicationName": "Testing Application",
  "email": "testing@test.com",
  "firstName": "Testing",
  "lastName": undefined,
  "preferences": Object {
    "notifications": Object {
      "channels": Array [
        Object {
          "enabled": true,
          "name": "daily-location-report",
          "options": Object {
            "hours": 9,
            "timezone": "America/Los_Angeles",
          },
        },
      ],
    },
  },
  "shifts": Array [
    Object {
      "_id": "5b480e6120d8f51322b0bfee",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0001",
    },
    Object {
      "_id": "60082e79dc704419ac763123",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0003",
    },
  ],
  "timezone": "America/Los_Angeles",
}
`;

exports[`helpers:getShiftDetails it returns filtered shifts when there are legacy location permissions 1`] = `
Object {
  "applicationName": "Testing Application",
  "email": "testing@test.com",
  "firstName": "Testing",
  "lastName": undefined,
  "preferences": Object {
    "notifications": Object {
      "channels": Array [
        Object {
          "enabled": true,
          "name": "daily-location-report",
          "options": Object {
            "hours": 9,
            "timezone": "America/Los_Angeles",
          },
        },
      ],
    },
  },
  "shifts": Array [
    Object {
      "_id": "5b480e6120d8f51322b0bfed",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0001",
    },
    Object {
      "_id": "60082e79dc704419ac763518",
      "files": Object {
        "pdf": Object {
          "path": "activity-reports/565fe59257c5d59e60cc35b3/1999/12/1999-12-31-1548194730622.pdf",
          "timestamp": 2019-12-31T18:00:00.000Z,
        },
      },
      "user": "565fe59257c5d59e60cc0002",
    },
  ],
  "timezone": "America/Los_Angeles",
}
`;
